model/src/main/scala/hmda/model/institution/Institution.scala

```diff
@@ -0,0 +1,53 @@
+package hmda.model.institution
+
+/**
+ * A financial institution, geared towards requirements for filing HMDA data.
+ */
+case class Institution(
+    id: Int,
+    name: String,
+    externalIds: Set[ExternalId],
+    agency: Agency,
+    institutionType: InstitutionType
+) {
+
+  val extIdsByType: Map[ExternalIdType, ExternalId] = externalIds.map(extId => (extId.idType, extId)).toMap
+
+  /**
+   * Derives the respondentId for a given Institution based on [[hmda.model.institution.Agency]] and [[hmda.model.institution.InstitutionType]],
+   * the rules for which can be found in section "1.4 - Respondent Identification Numbers for 2017 HMDA Filers" of the
+   * <a href="http://www.consumerfinance.gov/data-research/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf">2017 HMDA File Specifications</a>
+   */
+  def respondentId: Either[InvalidRespondentId, ExternalId] = {
+
+    institutionType.depositoryType match {
+      case None => Left(NoDepositoryTypeForInstitutionType(id, institutionType))
+      case Some(dt) =>
+
+        agency.externalIdTypes.get(dt) match {
+          case None => Left(UnsupportedDepositoryTypeByAgency(id, agency, dt))
+          case Some(extIdType) =>
+
+            extIdsByType.get(extIdType) match {
+              case None => Left(RequiredExternalIdNotPresent(id, extIdType))
+              case Some(extId) => Right(ExternalId(extId.id, extIdType))
+            }
+        }
+    }
```

Comment:
If we don't need the detailed error cases here (TBD), then this method can become much simpler, something like:
```
    val maybeRespId = for {
      depType <- institutionType.depositoryType
      extIdType <- agency.externalIdTypes.get(depType)
      extId <- extIdsByType.get(extIdType)
    } yield extId
    maybeRespId.toRight(RequiredExternalIdNotPresent(id, null)) // replace with whatever Left makes sense, or use Option
```
There may be slightly cleaner ways to write the last bit, but that's the basic idea.